lib/streamlit/UploadedFileManager.py

```diff
@@ -53,39 +44,45 @@ class UploadedFileManager(object):
     """
 
     def __init__(self):
-        self._files = {}
+        self._files_by_id = {}  # type: Dict[Tuple[str, str], UploadedFileList]
         self._files_lock = threading.Lock()
-        self.on_file_added = Signal(
-            doc="""Emitted when a file is added to the manager.
+        self.on_files_added = Signal(
+            doc="""Emitted when a file list is added to the manager.
 
             Parameters
             ----------
-            file : UploadedFile
-                The file that was added.
+            files : UploadedFileList
+                The file list that was added.
             """
         )
 
-    def add_file(self, file):
-        """Add a new file to the FileManager.
+    def add_files(self, session_id, widget_id, files):
+        """Add a list of files to the FileManager.
 
-        If another file with the same ID exists, it will be replaced with this
-        one.
+        If another list with the same (session_id, widget_id) key exists,
+        it will be replaced with this one.
 
-        The "on_file_added" Signal will be emitted after the file is added.
+        The "on_file_added" Signal will be emitted after the list is added.
 
         Parameters
         ----------
-        file : UploadedFile
-            The file to add.
+        session_id : str
+            The session ID of the report that owns the files.
+        widget_id : str
+            The widget ID of the FileUploader that created the files.
+        files : List[UploadedFile]
+            The files to add.
 
         """
+        file_list = UploadedFileList(
+            session_id=session_id, widget_id=widget_id, files=files
+        )
         with self._files_lock:
-            self._files[file.id] = file
-        self.on_file_added.send(file)
+            self._files_by_id[file_list.id] = file_list
+        self.on_files_added.send(file_list)
```

Comment:
We are ok, not locking here? Is it a problem if two threads send this signal concurrently? We would have potentially to process two lists on the listener side.

Reply:
`on_files_added` is a Signal, and Signals are themselves concurrency-safe (that is, you can fire signals off concurrently from multiple threads). The listeners themselves also need to be safe to call concurrently, of course - but our only listener to this signal (Server) meets this criteria.

Reply:
As far as each signal is executed on, but I think it is the way we do. So it is fine.